Connection and MAC pool module for Cisco UCS

dsoper2 opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
MAC pool configuration module for Cisco UCS Manager.  Includes requested changes from previous review of DNS module (https://github.com/ansible/ansible/pull/27117 ).  MAC pool being used for this PR to avoid dependency on ucsm_apis Python modules.  Subsequent UCS modules will follow this module's design once any needed changes are added.
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request
 - New Module Pull Request
 - Bugfix Pull Request
 - Docs Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
ucs_macpool
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.0 (ucsm bf8a747d06) last updated 2017/10/01 15:30:55 (GMT -500)
  config file = None
  configured module search path = [u'/Users/dsoper/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/dsoper/Documents/ansible/lib/ansible
  executable location = /Users/dsoper/Documents/ansible/bin/ansible
  python version = 2.7.12 (default, Oct 11 2016, 05:20:59) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.38)]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

Commit:
Initial commit for UcsConnection and ucs_macpool module.
Configures MAC address pools on UCS Manager.

Commit:
ansible-doc fixes

Commit:
PEP8 fixes

Commit:
pep8, pylint, and validate-modules fixes

Commit:
Correct indent issue introduced during pycodestyle cleanup

tima commented on lib/ansible/modules/remote_management/ucs/ucs_macpool.py:
```diff
@@ -0,0 +1,198 @@
+    mac_list:
+      - name: mac-A
+        from: from: 00:25:B5:00:66:00
+	to: 00:25:B5:00:67:F3
```
The indenting of this line and the next is off a space.

dsoper2 replied:
Hi Tim,
I have subsequent commits that address this and other PEP8/style issues so that all the CI checks are passing.

dagwieers commented on lib/ansible/module_utils/ucs.py:
```diff
@@ -0,0 +1,98 @@
+    hostname=dict(type='str', required=True),
+    username=dict(type='str', default='admin'),
+    password=dict(type='str', required=True, no_log=True),
+    port=dict(type='int', default=None),
```
This is not needed, as the default value defaults to None.

dsoper2 replied:
How would I avoid an error on use of port later in the file?  I confirmed this does fail if removed.

dagwieers commented on lib/ansible/module_utils/ucs.py:
```diff
@@ -0,0 +1,98 @@
+    port=dict(type='int', default=None),
+    use_ssl=dict(type='bool', default=True),
+    use_proxy=dict(type='bool', default=True),
+    proxy=dict(type='str', default=None),
```
This is not needed, as the default value defaults to None.

dagwieers commented on lib/ansible/modules/remote_management/ucs/ucs_macpool.py:
```diff
@@ -0,0 +1,201 @@
+
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
```
We now have a one-line GPLv3 license statement that is preferred.
```python
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

dagwieers commented on lib/ansible/utils/module_docs_fragments/ucs.py:
```diff
@@ -0,0 +1,77 @@
+    - UcsHandle object to interact with Cisco UCS Manager.
+      A single connection handle can be used across multiple modules rather than making a connection each time for each module.
+    type: UcsHandle
+    required: no
```
This is not needed. The default is that the parameter is not required.

dagwieers commented on lib/ansible/modules/remote_management/ucs/ucs_macpool.py:
```diff
@@ -0,0 +1,201 @@
+    module = _ansible_module_create()
+    conn = UcsConnection(module)
+    login_handle = conn.login()
+    result, err = setup(login_handle, module)
```
Why not put the real meat directly here in main() ?

I don't see a good reason to have 2 layers of abstraction (main -> setup -> setup_macpool) if it is unique to the module anyway.

dagwieers replied:
The advantage is that you can simple call module.fail_json there where is fails, without the need to pass everything back. I think it improves readability.

dagwieers commented on lib/ansible/modules/remote_management/ucs/ucs_macpool.py:
```diff
@@ -0,0 +1,201 @@
+
+
+def main():
+    module = _ansible_module_create()
```
I think this is confusing as it creates unneeded abstractions/indirections. I would keep this as simple as possible.

Why not just add parameters right here, and maybe comment what is MO-related and what is customized ?
It will make it much more clear what options are listed and helps keeping the documentation in sync as well.

*(This is just my opinion, if you disagree, that's fine too)*

dagwieers commented on lib/ansible/modules/remote_management/ucs/ucs_macpool.py:
```diff
@@ -0,0 +1,201 @@
+def main():
+    module = _ansible_module_create()
+    conn = UcsConnection(module)
+    login_handle = conn.login()
```
Why not make module a property from the connection object (maybe change that into a UCS object instead ?).
Make it automatically log on if there's no handle already, and keep the handle of the UCS object as well.

Then you can simply call the setup_macpool function, and it would have access to both the module as well as the login handle, making the module much cleaner.

dsoper2 replied:
I've changed the main structure per the other comments to avoid the unneeded function calls.  Because the python SDK (and all of our other modules that will be added after this) use the login/logout directly my thought is to leave that sort of code flow.

dagwieers commented on lib/ansible/modules/remote_management/ucs/ucs_macpool.py:
```diff
@@ -0,0 +1,201 @@
+    conn = UcsConnection(module)
+    login_handle = conn.login()
+    result, err = setup(login_handle, module)
+    conn.logout()
```
This could be part of the setup_macpool function in this case.

Although it would also be nice if somehow we could register calls like this automatically so that whenever the module breaks down, it would automatically disconnect. I know this is not yet possible today, but would be very useful. Especially since UCS is very sensitive to open/unused handles.

dagwieers commented on lib/ansible/module_utils/ucs.py:
```diff
@@ -0,0 +1,98 @@
+
+        from ucsmsdk.ucshandle import UcsHandle
+        results = {}
+        try:
```
We try to avoid large try-except blocks and prefer if this is done for specific calls only.
In this case I assume the try-except is only useful for UcsHandle, so it would be useful at some point in the future to provide better error-message if that's needed.

*(Especially for network/connection related issues where you could translate a Connection Refused to something with more context to the user)*

dsoper2 replied:
Did you have thoughts on better structuring to handle use_proxy and proxy together?  I didn't have an idea for a better way to handle those variables along with the use of UcsHandle proxy for the 3 cases it supports:
- proxy=None uses system defined proxy
- proxy={} uses no proxy
- proxy=str uses the user defined proxy (and overrides system proxy setting)

dagwieers replied:
So other modules do it like this:
- `proxy=None` (default) and `use_proxy=yes` (default) should be using the system defined proxy
- `proxy=None` (default) and `use_proxy=no` (user) should not be using a proxy
- `proxy=value` and `use_proxy=yes` (default) should use the provided proxy
Next to this often `proxy_username` and `proxy_password` are configurable for proxy authentication.

dagwieers reviewed (changes requested):
Same questions and small changes.

dagwieers commented:
The use of **login_handle** as a parameter that can be passed on is something more modules need (even ACI, IMC, but most REST-based APIs) and should be something that is standardized or even works automatically (values are automatically passed on through hostvars, and automatically renewed when expired).

**Update:** I added a proposal for this at https://github.com/ansible/proposals/issues/81

Commit:
Simplified module arugment setup.
Placed all code in main to avoid multiple calls and arg passing.

Commit:
module_utils/ucs changed to UCSModule which now handles login/logout directly
login_handle removed from module.params
doc updates on mac_list params and change to first_addr/last_addr for mac blocks
checking of all mac params

dagwieers commented:
We may want to move the ucs library into a sub-directory named remote_management, just like networking is now moving to this scheme for good reasons.

Commit:
Move module_utils to remote_management/ucs
Fix validate-modules issue with docs

Commit:
UCS MAC pool integration tests
Fixed issues with MAC pool descr and address range params

movinalot reviewed (approved):
All looks good

dsoper2 commented:
@dagwieers I have made the changes you requested and now have 2 additional approvals.  Anything else needed for submitting upstream?

dagwieers reviewed (approved):
LGTM
